### ISSUE TITLE
Are groups or sets of variables a good way to handle wide (many variables) in R-Instat

### ISSUE THREAD

**Issue by rdstern** [IDEMSInternational/R-Instat#4009]
- [X] Create dialog for make/edit/delete sets
- [X] Write R methods for make/edit/delete sets
- [ ] Link dialog to R methods
- [ ] Think out dialogs for applying sets

For the procurement work we have a long (185, 000 cases) and wide (192 variables) data set.  It is an excellent example, and we will be making some simple videos explaining how to "tame" this data set.

Note that our default window only shows 1000 cases by 30 variables.  This can be changed, but that is not the point.  If we want to tempt people to "look at data" then 1000 by 30 i.e. 30,000 values is already quite large.

We can already use our powerful filter system to "tame" the number of rows.  For example the data are from over 200 countries.  We could use the filter to look just at one country.  For example there are just over 1000 records for Kenya.  

How do we "tame" the columns.  Just for reference we have a really big way to tame the columns already.  Often wide data (lots of variables) are actually from multiple levels.  In this case some of the variables are at the **project level**, while others are at the **contract level within each project**.  We already have the facility to look at multiple (linked) sheets and that is really important.

In addition, however, we need to "tame" the variables within a sheet.  I think of these as groups of variables.  SPSS calls them "sets", possibly because a variable can belong to multiple sets.  What should we call them?  This is all part of our metadata and I need to check whether this means anything in R.

Here is the SPSS dialogue to complete the background.  In the comment below I suggest the structure for the R-Instat dialogue.
![image](https://user-images.githubusercontent.com/11226469/31267622-b6d522d4-aa70-11e7-82e5-feecab1b3183.png)

**Comment by rdstern** [comment]
So "our" dialogue will be really easy.  The name of the dialogue could be **Define sets of variables**.

1. At the top there will be 3 buttons Make, Edit, Delete.
2. **To define a set** we have our usual Data selector and a multiple receiver.
3. Then there is a field to give a name to the set.  This could be labelled "Set Name:"
That's all really?
4. **To edit a set** there is the usual selector, but it now shows only the sets that have been defined.
5. There is a single receiver to choose a set.
6. Then there is a multiple receiver that contains all the variables in the chosen set.  
7. The field for Set Name is still there with the old name.  If we don't change that name, then the set is edited.  If we do change it, then the original set remains and the new name is added.
8.  To delete a set, there is the same selector as for editing.  There is a multiple receiver.  
9. There is no field to receive the name.  There could be the same sort of warning as for deleting columns or data frames.

**Comment by maxwellfundi** [comment]
@Patowhiz I think could be a good task for you today. We can discuss before you take it on.

**Comment by rdstern** [comment]
Great.  Danny is of the view that the dialogue would go into the Prepare > Data Frame menu.  It could be in the same section as Column Structure. Maybe just after that and above the line.

**Comment by rdstern** [comment]
I am less clear how we want to proceed on the remaining dialogues.  To help our discussion I show the situation in SPSS.  First the menu, where there are 3 items:
![image](https://user-images.githubusercontent.com/11226469/31269618-15d69f6c-aa79-11e7-8b2f-deeef2972267.png)

Note the icons in SPSS within the menu for **Define, Use and Show All** variables.

Then the second dialogue in SPSS where you define the use of one or more of the sets of variables:  
![image](https://user-images.githubusercontent.com/11226469/31269735-72f2948a-aa79-11e7-90e8-ca48d3fe3c55.png)

I have shown the toolbar here where it states that 7 out of the 25 variables are being show.  And these items are on the toolbar.

In our case at the bottom of the grid we already show the stets of the filter, so that is the obvious place to show the number of columns being displayed.  
I like the idea of a drop down with the sets so that we can choose.  Might this even be the same dialogue as above with a 4th button for **Use**.  That would be quite easy.  Or it could be a Second Dialogue called **Use Sets**.
If it is a single dialogue then the Menu item would be "Define/Use Sets".  And the top of the dialogue would say "Define and Use Sets of Variables".  Either way the use sets would (like SPSS) not need the data selector.  It would just have a small grid with the defined sets and the ability to tick one or more.  There should be an option to see all the variables.  perhaps, like SPSS one entry could always be All Variables.

What do we do to remove the selection of variables.  I notice that we have Remove Current Filter on the right-click of both columns and rows.  But I don't see the menu option to remove a filter.  Interesting.  Some discussion is needed here.  Are there instances where users would not use the right-click?

**Comment by dannyparsons** [comment]
You can apply the "no filter" in the dialog and this will remove the currently filter, but yeah it's not obvious.

Not sure if Create, Edit and Delete Sets should be all in the same dialog, whether that's easier or more confusing for the users, but happy with it for now.

I would suggest that in delete sets you do still have the multiple selector for columns, but it's disabled so you can see the columns in the set you are going to delete.

These sets will just be stored in the metadata but I think it's worth having separate methods for them in the instat object which I can write.

**Comment by rdstern** [comment]
You are right that the no filter option is not obvious.  However it does assist in avoiding haveing an extra menu item - except on the right-click.
When I compare the R-Instat menus to SPSS, Genstat, Minitab a standout feature is the length (and to me sometimes confusing nature) of the menus.

So I am happy that R-Instat's menus are relatively short.
We must not compromise on the fact they remain simple.  I think our new radio buttons help here.  We are going to be adding items as we add more features.  So, on balance, when a single dialogue (like Filter) can achieve these multiple objectives, then I prefer this to a proliferation of menu items/dialogues

**Comment by rdstern** [comment]
Further item on this general area.  SPSS applies the visible sets also to the showing of the column metadata.  I like that.  There it looks like an ordinary filter, so  the colour of the column that gives the variable number could change (as we do for the row number in the ordinary filter).  And we keep the original variable number, just as we do in the data grid when we apply a filter.

**Comment by maxwellfundi** [comment]
@rdstern  just to confirm that the top buttons are our usual  radiobuttons

**Comment by dannyparsons** [comment]
Yes

**Comment by rdstern** [comment]
Yes indeed.  Our new-style improved radio buttons that David was so keen to have (instead of the old fashioned ones I was used to).  Just occasionally his ideas are good ones!

**Comment by rdstern** [comment]
Further items to consider - though they could come later.
1. Could we have a dialogue - perhaps the existing **Prepare > Data Frame > Colour by Property** that could include the option to colour any group.  It is a property after all.  I have never used it so far!
2. If we were to make a subset - through our filter for example - might the groups be copied through.  Might this happen automatically?
3. We already have a dialogue in the menu we will use here, which is **Prepare > Data Frame > Column Structure**.  This seems to me like a special case of these sets, though where the 3 sets are mutually exclusive.  That dialogue would allow (I assume) implicit changes in the 3 groups, but there is no delete.  Should the dialogues be combined?

**Comment by maxwellfundi** [comment]
@Patowhiz  can you get this to the next level now? Its needed for tomorrows release

**Comment by Patowhiz** [comment]
@maxwellfundi , @dannyparsons. I just uploaded the code for this dialogue for review. I still need help on how we can save the created variable sets and make it accessible in other dialogues. Currently the created variable sets are only accessible via this form ONLY.

**Comment by Patowhiz** [comment]
@maxwellfundi , @dannyparsons. I just uploaded the code for this dialogue for review. I still need help on how we can save the created variable sets and make it accessible in other dialogues. Currently the created variable sets are only accessible via this form ONLY.

**Comment by dannyparsons** [comment]
Yes, we then need ways of using the sets, probably in the Data Options but also another dialog. That can be a separate pull request. Can you get someone to review your pull request so it can be merged soon?

**Comment by Patowhiz** [comment]
@dannyparsons yeah.

**Comment by Patowhiz** [comment]
Great we now need to discuss on how the variable sets would be accessible to the dialogs, in particular to those dialogs selectors. @dannyparsons suggested the data options, what are your thoughts @rdstern ?

**Comment by dannyparsons** [comment]
It should be in the Data Options - and we have a tab for Columns which is where it should go. The design of that will need some thought.

I think Roger also suggested on the right click in the selector. There could be a "Sets" item which has sub items to let you check/uncheck a group to make those variables visible/invisible in the selector.

**Comment by lilyclements** [comment]
I note in #6426 we say that that issue "builds on, and supersedes topics mentioned in https://github.com/africanmathsinitiative/R-Instat/issues/4372, in https://github.com/africanmathsinitiative/R-Instat/issues/4957 and also in https://github.com/africanmathsinitiative/R-Instat/issues/6263 and in https://github.com/africanmathsinitiative/R-Instat/issues/4009. "

Am I correct in saying that this PR can be closed? (alongside with any of those other three that are open)

**Comment by lilyclements** [comment]
@rdstern I note in #6426 we say that that issue "builds on, and supersedes topics mentioned in https://github.com/africanmathsinitiative/R-Instat/issues/4372, in https://github.com/africanmathsinitiative/R-Instat/issues/4957 and also in https://github.com/africanmathsinitiative/R-Instat/issues/6263 and in https://github.com/africanmathsinitiative/R-Instat/issues/4009. "

Am I correct in saying that this PR can be closed? (alongside with any of those other three that are open)
